src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/InteropServices/PlatformCompatabilityAnalyzer.cs

```diff
@@ -0,0 +1,847 @@
+                }
+                else if (attribute.UnsupportedSecond != null)
+                {
+                    ReportUnsupportedDiagnostic(operation, context, operationName, platformName, VersionToString(attribute.UnsupportedSecond));
+                }
+
+                if (attribute.Obsoleted != null)
+                {
+                    context.ReportDiagnostic(operation.CreateDiagnostic(ObsoleteOsRule, operationName, platformName, attribute.Obsoleted)); ;
+                }
+            }
+        }
+
+        private static void ReportSupportedDiagnostic(IOperation operation, OperationBlockAnalysisContext context, string name, string platformName, string? version = null) =>
+            context.ReportDiagnostic(version == null ? operation.CreateDiagnostic(RequiredOsRule, name, platformName) :
+                operation.CreateDiagnostic(RequiredOsVersionRule, name, platformName, version));
+
+        private static void ReportUnsupportedDiagnostic(IOperation operation, OperationBlockAnalysisContext context, string name, string platformName, string? version = null) =>
+            context.ReportDiagnostic(version == null ? operation.CreateDiagnostic(UnsupportedOsRule, name, platformName) :
+                operation.CreateDiagnostic(UnsupportedOsVersionRule, name, platformName, version));
+
+        private static string? VersionToString(Version version) => IsEmptyVersion(version) ? null : version.ToString();
+
+        private static ISymbol? GetOperationSymbol(IOperation operation)
+            => operation switch
+            {
+                IInvocationOperation iOperation => iOperation.TargetMethod,
+                IObjectCreationOperation cOperation => cOperation.Constructor,
+                IFieldReferenceOperation fOperation => IsWithinConditionalOperation(fOperation) ? null : fOperation.Field,
+                IMemberReferenceOperation mOperation => mOperation.Member,
+                _ => null,
+            };
+
+        private static void AnalyzeOperation(IOperation operation, OperationAnalysisContext context,
+            PooledConcurrentDictionary<IOperation, SmallDictionary<string, PlatformAttributes>> platformSpecificOperations,
+            PooledConcurrentDictionary<ISymbol, SmallDictionary<string, PlatformAttributes>> platformSpecificMembers)
+        {
+            var symbol = GetOperationSymbol(operation);
+
+            if (symbol == null)
+            {
+                return;
+            }
+
+            if (!platformSpecificMembers.TryGetValue(symbol.OriginalDefinition, out var operationAttributes) &&
+                TryFindPlatformAttributesApplied(symbol.GetAttributes(), symbol.OriginalDefinition.ContainingSymbol, out operationAttributes))
+            {
+                platformSpecificMembers.TryAdd(symbol.OriginalDefinition, operationAttributes);
+            }
```

Comment:
I think we should add to the map even when invoked symbol has no platform applied attributes. Otherwise, you will end up recomputing the attributes every time for any non-platform specific symbol invoked large number of times.
```suggestion
            if (!platformSpecificMembers.TryGetValue(symbol.OriginalDefinition, out var operationAttributes))
            {
                var hasAttributes = TryFindPlatformAttributesApplied(symbol.GetAttributes(), symbol.OriginalDefinition.ContainingSymbol, out operationAttributes);
                Debug.Assert(hasAttributes || operationAttributes == null);
                Debug.Assert(!hasAttributes || operationAttributes != null && operationAttributes.Any());

                platformSpecificMembers.TryAdd(symbol.OriginalDefinition, operationAttributes);
            }
```